Mathlib/Data/Finset/Sym.lean

```diff
@@ -0,0 +1,269 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+
+! This file was ported from Lean 3 source module data.finset.sym
+! leanprover-community/mathlib commit 98e83c3d541c77cdb7da20d79611a780ff8e7d90
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Finset.Lattice
+import Mathlib.Data.Fintype.Prod
+import Mathlib.Data.Fintype.Vector
+import Mathlib.Data.Sym.Sym2
+
+/-!
+# Symmetric powers of a finset
+
+This file defines the symmetric powers of a finset as `Finset (Sym α n)` and `Finset (Sym2 α)`.
+
+## Main declarations
+
+* `Finset.sym`: The symmetric power of a finset. `s.sym n` is all the multisets of cardinality `n`
+  whose elements are in `s`.
+* `Finset.sym2`: The symmetric square of a finset. `s.sym2` is all the pairs whose elements are in
+  `s`.
+
+## TODO
+
+`Finset.sym` forms a Galois connection between `Finset α` and `Finset (Sym α n)`. Similar for
+`Finset.sym2`.
+-/
+
+namespace Finset
+
+variable {α : Type _} [DecidableEq α] {s t : Finset α} {a b : α}
+
+theorem isDiag_mk'_of_mem_diag {a : α × α} (h : a ∈ s.diag) : Sym2.IsDiag ⟦a⟧ :=
+  (Sym2.isDiag_iff_proj_eq _).2 (mem_diag.1 h).2
+#align finset.is_diag_mk_of_mem_diag Finset.isDiag_mk'_of_mem_diag
+
+theorem not_isDiag_mk'_of_mem_offDiag {a : α × α} (h : a ∈ s.offDiag) : ¬Sym2.IsDiag ⟦a⟧ := by
+  rw [Sym2.isDiag_iff_proj_eq]
+  exact (mem_offDiag.1 h).2.2
+#align finset.not_is_diag_mk_of_mem_off_diag Finset.not_isDiag_mk'_of_mem_offDiag
+
+section Sym2
+
+variable {m : Sym2 α}
+
+/-- Lifts a finset to `Sym2 α`. `s.sym2` is the finset of all pairs with elements in `s`. -/
+-- Porting note: changed ×ˢ to xᶠ
+protected def sym2 (s : Finset α) : Finset (Sym2 α) := (s ×ᶠ s).image Quotient.mk'
+
+
+#align finset.sym2 Finset.sym2
+
+@[simp]
+theorem mem_sym2_iff : m ∈ s.sym2 ↔ ∀ a ∈ m, a ∈ s := by
+  refine
+    mem_image.trans
+      ⟨?_, fun h ↦ ⟨m.out, mem_product.2 ⟨h _ m.out_fst_mem, h _ m.out_snd_mem⟩, m.out_eq⟩⟩
+  rintro ⟨⟨a, b⟩, h, rfl⟩
+  rw [Quotient.mk', @Sym2.ball _ (fun x ↦ x ∈ s)]
+  rwa [mem_product] at h
+#align finset.mem_sym2_iff Finset.mem_sym2_iff
+
+theorem mk'_mem_sym2_iff : ⟦(a, b)⟧ ∈ s.sym2 ↔ a ∈ s ∧ b ∈ s := by rw [mem_sym2_iff, Sym2.ball]
+#align finset.mk_mem_sym2_iff Finset.mk'_mem_sym2_iff
+
+@[simp]
+theorem sym2_empty : (∅ : Finset α).sym2 = ∅ := rfl
+#align finset.sym2_empty Finset.sym2_empty
+
+@[simp]
+theorem sym2_eq_empty : s.sym2 = ∅ ↔ s = ∅ := by
+  rw [Finset.sym2, image_eq_empty, product_eq_empty, or_self_iff]
+#align finset.sym2_eq_empty Finset.sym2_eq_empty
+
+@[simp]
+theorem sym2_nonempty : s.sym2.Nonempty ↔ s.Nonempty := by
+  rw [Finset.sym2, Nonempty.image_iff, nonempty_product, and_self_iff]
+#align finset.sym2_nonempty Finset.sym2_nonempty
+
+alias sym2_nonempty ↔ _ nonempty.sym2
+#align finset.nonempty.sym2 Finset.nonempty.sym2
+
+-- Porting note: attribute does not exist
+-- attribute [protected] nonempty.sym2
```

Comment:
This is problematic; the intent is to create a protected alias. If we can't do it this way, perhaps there's other syntax?

Reply:
This is discussed here: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/alias.20.26.20protected but I don't think a solution has been found